### PR TITLE
WIF support design

### PR DESCRIPTION
**Pull request by kaovilai**
Closes https://github.com/openshift/oadp-operator/issues/1047

Implementation: https://github.com/openshift/oadp-operator/pull/1112

### CODE REVIEW

File: docs/design/gcp-wif-support.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    echo 'apiVersion: cloudcredential.openshift.io/v1
+    kind: CredentialsRequest
+    metadata:
+      name: oadp-operator-credentials
+      namespace: openshift-cloud-credential-operator
+    spec:
+      providerSpec:
+        apiVersion: cloudcredential.openshift.io/v1
+        kind: GCPProviderSpec
+        predefinedRoles:
```
Comment (kaovilai) [inline docs/design/gcp-wif-support.md:66]: These are permissions
```suggestion
        permissions:
```
need to validate some perms

File: docs/design/gcp-wif-support_design.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        - storage.objects.delete
+        - storage.objects.get
+        - storage.objects.list
+        - iam.serviceAccounts.signBlob
+        skipServiceCheck: true
+      secretRef:
+        name: cloud-credentials-gcp
+        namespace: openshift-adp
+      serviceAccountNames:
+      - velero
```
Comment (shubham-pampattiwar) [inline docs/design/gcp-wif-support_design.md:74]: Shouldn't this be operator SA `openshift-adp-controller-manager` ?

Comment (kaovilai) [reply docs/design/gcp-wif-support_design.md:74]: nah. this is for velero SA to use.openshift-adp-controller-manager sa do not actually access or create data to bucket/snapshots.

File: docs/design/gcp-wif-support_design.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+In currently released versions of OADP, the only way to authenticate to GCP is via a service account.
+This is not ideal for customers who are using GCP's WIF feature to authenticate to GCP.
+This proposal aims to add support for WIF to OADP.
+
+## Goals
+- GCP WIF support for OADP and Velero for backup and restore of applications backed by GCP resources.
+- Using OpenShift's Cloud Credentials Operator to generate a short-lived token for authentication to GCP.
+
+## Non Goals
+- ImageStreamTag backup and restore (follow up design/implementation)
```
Comment (shawn-hurley) [inline docs/design/gcp-wif-support_design.md:23]: Can you explain, for my edification, why this is a non-goal?

Comment (shawn-hurley) [reply docs/design/gcp-wif-support_design.md:23]: I am trying to understand why we should consider these discrete parts of getting this working. IMO getting WIF support means that it works for all the things that we currently get when using the service account.

Comment (sseago) [reply docs/design/gcp-wif-support_design.md:23]: @shawn-hurley *If* we have a use case that matters where local image copy is not required, then getting non-image bits working sooner might be advantageous. On the other hand, if the most important use cases require local image backup, then we probably want it all here.

@shubham-pampattiwar  do you happen to know how important a use case the image backup one is? Is this something relevant to more than 50% of users, or is it more like 10% or lower? It's mainly relevant for use cases where images are manually uploaded to the internal registry -- for Build/BuildConfig-driven images, they'll be rebuilt from scratch on restore of BuildConfig anyway.

Comment (shubham-pampattiwar) [reply docs/design/gcp-wif-support_design.md:23]: @sseago To be perfectly I am unsure of the image backup use case relevance number, but maybe, maybe its lets than 50% of the users.

File: docs/design/gcp-wif-support_design.md (hunk 2 above)
Comment (shubham-pampattiwar) [reply docs/design/gcp-wif-support_design.md:74]: oh ok, makes sense.

File: docs/design/gcp-wif-support_design.md (hunk 3 above)
Comment (weshayutin) [reply docs/design/gcp-wif-support_design.md:23]: Hard for me to tell how many customers find image backup important.  TBH backing up images seems like an anti-pattern and should be discouraged, however the customer is always right :)  To better answer the question, I find it difficult to prioritize image backup.

Comment (weshayutin) [reply docs/design/gcp-wif-support_design.md:23]: Hard for me to tell how many customers find image backup important.  TBH backing up images seems like an anti-pattern and should be discouraged, however the customer is always right :)  To better answer the question, I find it difficult to prioritize image backup.

**Comment by kaovilai** [comment]
> Am I missing why OADP would not create the request for itself?

We're just following. https://github.com/abutcher/gcp-wif-example/tree/main to start with.
You will note that it uses `ccoctl` to generate a service account by using `gcloud` CLI directly (try `--dry-run`).

If automation is possible, I doubt [this page](https://docs.openshift.com/container-platform/4.13/security/cert_manager_operator/cert-manager-authenticate-gcp.html) would exists in the ocp Documentation.

**Comment by kaovilai** [comment]
When https://issues.redhat.com/browse/OCPSTRAT-95 is complete, we could come back and look at _creating the request for itself_.

File: docs/design/gcp-wif-support_design.md (hunk 3 above)
Comment (shawn-hurley) [reply docs/design/gcp-wif-support_design.md:23]: I still firmly believe we must understand the complete design and the impacts before we decide which parts of work will be available in this release. 

I am not asking that sign up for all the work in the design, just that we think it about and can state what changes we do need for them.

Comment (kaovilai) [reply docs/design/gcp-wif-support_design.md:23]: So state changes needed to implement non-goals and that we are not doing it yet?
Or move it to goals and mark _to be completed later_?

Comment (shawn-hurley) [reply docs/design/gcp-wif-support_design.md:23]: Whichever works for you works for me, I just think that we need to discuss it.

Comment (kaovilai) [reply docs/design/gcp-wif-support_design.md:23]: Moving into goals as Workload Identity work https://github.com/distribution/distribution/commit/32b33ab6bd6595173762350993093eb2fce9f83b is already in [release-4.14](https://github.com/distribution/distribution/compare/main...openshift:docker-distribution:release-4.14) of docker-distribution so distribution repo work is already eliminated

**Comment by kaovilai** [comment]
> Why OADP would not create the request for itself?

There is no _CloudCredentialOperator-based flow for OLM-managed operators and WIF_ documentation yet.

Only AWS STS and Azure Identity standardized flow are available afaik.

@shawn-hurley @shubham-pampattiwar

**Comment by kaovilai** [comment]
> Why OADP would not create the request for itself?

There is no _CloudCredentialOperator-based flow for OLM-managed operators and WIF_ documentation yet.

Only AWS STS and Azure Identity.

https://issues.redhat.com/browse/OCPSTRAT-95

**Comment by weshayutin** [comment]
Thank you @kaovilai well done!

File: docs/design/gcp-wif-support_design.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        - compute.zones.get
+        - storage.objects.create
+        - storage.objects.delete
+        - storage.objects.get
+        - storage.objects.list
+        - iam.serviceAccounts.signBlob
+        skipServiceCheck: true
+      secretRef:
+        name: cloud-credentials-gcp
+        namespace: openshift-adp
```
Comment (shubham-pampattiwar) [inline docs/design/gcp-wif-support_design.md:72]: NIT: s/openshift-adp/<OPERATOR_INSTALL_NS>

File: docs/design/gcp-wif-support_design.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+This aids in review of the document as changes to a line are not obscured by the reflowing those changes caused and has a side effect of avoiding debate about one or two space after a period.
+
+_Note_: The name of the file should follow the name pattern `<short meaningful words joined by '-'>_design.md`, e.g:
+`listener-design.md`. -->
+
+## Abstract
+Support Google Cloud's WIF (Workload Identity Federation) for OADP.
+
+## Background
+In currently released versions of OADP, the only way to authenticate to GCP is via a service account.
```
Comment (kaovilai) [inline docs/design/gcp-wif-support_design.md:14]:
```suggestion
In currently released versions of OADP, the only way to authenticate to GCP is via a long lived service account credentials.
```

File: docs/design/gcp-wif-support_design.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+This is not ideal for customers who are using GCP's WIF feature to authenticate to GCP.
```
Comment (kaovilai) [inline docs/design/gcp-wif-support_design.md:15]:
```suggestion
This is not ideal for customers who are using GCP's WIF ([Workload Identity](https://cloud.google.com/kubernetes-engine/docs/concepts/workload-identity)) feature to authenticate to GCP.
```

File: docs/design/gcp-wif-support_design.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+		veleroContainer.VolumeMounts = append(veleroContainer.VolumeMounts,
+			corev1.VolumeMount{
+				Name:      "bound-sa-token",
+				MountPath: "/var/run/secrets/openshift/serviceaccount",
+				ReadOnly:  true,
+			}),
+```
+
+<!-- ## Open Issues
+A discussion of issues relating to this proposal for which the author does not know the solution. This section may be omitted if there are none. -->
```
Comment (shubham-pampattiwar) [inline docs/design/gcp-wif-support_design.md:155]: Maybe add the standardized OLM workflow for GCP WIF issue here ?

**Review by shubham-pampattiwar — APPROVED** [review]

File: docs/design/gcp-wif-support_design.md (hunk 7 above)
Comment (kaovilai) [reply docs/design/gcp-wif-support_design.md:155]: Cool. Thanks!

File: docs/design/gcp-wif-support_design.md (hunk 4 above)
Comment (kaovilai) [reply docs/design/gcp-wif-support_design.md:72]:
```suggestion
        namespace: <OPERATOR_INSTALL_NS>
```

File: docs/design/gcp-wif-support_design.md (hunk 7 above)
Comment (kaovilai) [inline docs/design/gcp-wif-support_design.md:155]:
```suggestion
## Open Issues
<!-- A discussion of issues relating to this proposal for which the author does not know the solution. This section may be omitted if there are none. -->
- [Standardized update flow for OLM-managed operators leveraging short-lived token authentication](https://issues.redhat.com/browse/OCPSTRAT-95) do not yet have support for WIF. We will have to follow up with this work later.

```

**Review by sseago — APPROVED** [review]

**Review by weshayutin — APPROVED** [review]

**Pull request merged**